airrohr-firmware/mhz19.cpp

```diff
@@ -0,0 +1,150 @@
+ /***********************************************************************
+ *    MH-Z19 CO2 sensor driver                                          *
+ *    Copyright (C) 2017-2019 Michael Dreher <michael (at) 5dot1.de>    *
+ *                                                                      *
+ * This program is free software: you can redistribute it and/or modify *
+ * it under the terms of the GNU General Public License as published by *
+ * the Free Software Foundation, either version 3 of the License, or    *
+ * (at your option) any later version.                                  *
+ *                                                                      *
+ * This program is distributed in the hope that it will be useful,      *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
+ * GNU General Public License for more details.                         *
+ *                                                                      *
+ * You should have received a copy of the GNU General Public License    *
+ * along with this program. If not, see <http://www.gnu.org/licenses/>. *
+ ***********************************************************************/
+
+#include "mhz19.h"
+
+CMHZ19Sensor::CMHZ19Sensor(Stream& serialPort) :
+	m_pSerialPort(&serialPort),
+	ppmMaxValue(0xffff)
+{
+}
+
+/*
+set the measuring range to 1000, 2000, 3000 or 5000
+ppm range for sensor MH-Z19 is set to 2000 when sensor is delivered
+*/
+bool CMHZ19Sensor::SetRange(uint16_t range)
+{
+  if((range % 1000) == 0
+    && range >= 1000
+    && range <= 5000
+  )
+  {
+    uint8_t cmd[packetLen] = {0xFF, 0x01, 0x99, 0x00, 0x00, 0x00, (uint8_t)((range >> 8) & 0xff), (uint8_t)(range & 0xff), 0x00}; // 2nd uint8_t is the sensor number
+    cmd[packetLen - 1] = CalcCheckSum(cmd);
+    m_pSerialPort->write(cmd, sizeof(cmd));
+    // TODO: check return value if any
+    // consume any response in the receive buffer
+    delay(200);
+    while(m_pSerialPort->available())
+      m_pSerialPort->read();
+
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
+
+uint16_t CMHZ19Sensor::GetRange()
```

Comment:
The whole getrange logic doesn't seem to be used? Is this expected?

Reply:
I first thought it would have an effect on the accuracy, but the only reference in the docu is for the scaling of the PWM value, not for the UART output.

Removed range logic (hidden behind #ifdef)